mitodl/ocw-studio
# import menus.yaml files as navmenus and update test data

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/430
Closes https://github.com/mitodl/ocw-studio/issues/410

#### What's this PR do?
A number of changes have been made to the way `ocw-to-hugo` outputs data and the `ocw_import` code here in `ocw-studio` needs to be updated to reflect those changes.  Firstly, the left navigation menu `leftnav` now has its entries stored in a menu specific configuration file stored in each course under `config/_default/menus.yaml` instead of the front matter of each page that had a nav entry.  The primary purpose of this PR is to correct that by removing the code that pulls nav entries from front matter and instead sources them from the `menus.yaml` file.  While working on this, a few other issues were uncovered that were addressed here.  The parent / child relationship of `WebsiteContent` objects was being determined from these front matter based menu entries, and that has been moved over to simply use a `parent_uid` property introduced in https://github.com/mitodl/ocw-to-hugo/pull/320.  Also, `WebsiteContent` objects created by the import are not currently visible in the UI because pages did not have `is_page_content` set to `True`.  This has been corrected so imported pages properly show up in the UI.  Aside from that, the last change was updating the `ocw-to-hugo` test data to the latest version and updating the tests.

#### How should this be manually tested?
Proper manual testing of this PR requires a release of https://github.com/mitodl/ocw-to-hugo/pull/320.  Once this is complete:
 - Set up `ocw-studio` to run locally and make sure S3 access is properly configured
 - Run `docker-compose run web python manage.py import_ocw_course_sites --bucket ocw-to-hugo-output-production --filter 15-401`
 - Browse to the imported site locally at http://localhost:8043/sites/15-401-finance-theory-i-fall-2008
 - Ensure that pages were imported in the Page section and they are visible in the UI
 - Ensure that the menu was imported and items under the Video Lectures and Slides section are indented to indicate that they are children of that section

#### Any background context you want to provide?
 One test in particular was problematic, `test_import_ocw2hugo_course_bad_date`.  After updating the test data, there are more `WebsiteContent` objects that get imported and it revealed a problem with the way this test was implemented.  The test was running `mocker.patch` in a way that caused unintentional side effects.  In `api.py`, we were assigning `dateutil.parser` to a variable like:

`from dateutil import parser as dateparser`

...and in the test we were patching it like:

```
    mocker.patch(
        "ocw_import.api.dateparser.parse", side_effect=["2021-01-01", ValueError()]
    )
```

When there were less `WebsiteContent` objects being imported this was fine, but when the test data was updated and more items were added the side effect became apparent.  Even in the old data, the parse function should have only been called once and the array of side effects shouldn't have been necessary.  Calling `mocker.patch` and passing in `ocw_import.api.dateparser.parse` here does not limit the patch to just the locally assigned `dateparser` variable in `api.py` as one might think.  Instead *any* call to `dateutil.parser.parse` is patched.  This means that updates to external dependencies outside of our code even could potentially add extra calls to that that would get patched, resulting in a `StopIteration` error.  Instead of patching the function in this way, I made a small wrapper function in `api.py` for parsing dates and patched that function instead.  Now the patch is only called once (as expected) and the test behaves as it should.
